kernel: make `k_msgq_put` become scheduler point

TaiJuWu opened the pull request:
Fix: https://github.com/zephyrproject-rtos/zephyr/issues/79617

TaiJuWu committed:
kernel: `k_msgq_put` should trigger reschedule

Make `k_msgq_put` as scheduler point to resolve
high priority thread can not preempt low priority
thread.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

andyross commented on kernel/msg_q.c:
```diff
@@ -178,7 +182,11 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
+		z_reschedule(&msgq->lock, key);
+	} else {
+		k_spin_unlock(&msgq->lock, key);
+	}
```
FWIW I'm not a fan of this pattern, which is common but has been preserved only for compatibility.  Basically what this is saying is that "k_msgq_put() is a scheduling point if there was a thread waiting, **but not if there wasn't**", which is confusing to document and reason about.  (The difference in behavior being whether or not it will allow a third runnable higher priority thread to run or not).

Ideally an operation would be either a scheduling point or not, without this kind of logic.  Again we don't currently do that in many places because of history, and because changing scheduler behavior is often perceived as a compatibility bug.    But as we evolve it should be in that direction IMHO.

TaiJuWu replied:
Thanks for your explanation. It is very useful for me.

andyross reviewed (commented):
No complaints in general to making this a scheduling point, but:

1. Please correct the commit message.  This isn't changing wakeup behavior, just whether or not a reschedule happens.
2. Flip the order of the two commits to avoid a bisection glitch (the test fails in the current tree, right?)
3. Add documentation to k_msgq_put() to make it clear that the call will reschedule.
4. See nitpick about the "predicated reschedule()" anti-pattern that IMHO we don't want to perpetuate.

andyross reviewed (dismissed):
I'm out of complaints, thanks!

nashif commented on tests/integration/kernel/src/main.c:
```diff
@@ -302,6 +306,38 @@ ZTEST(kernel, test_mutex_usage)
 	k_thread_abort(tid2);
 }
 
+static void high_prio_main(void *param1, void *param2, void *param3)
```
msgq tests should go into `tests/kernel/msgq/`

TaiJuWu replied:
Hi @nashif ,
This test include two behavior `poll`, `message` and require run on only 1 cpu so I think it is better here than `msgq`.
WDYT?

nashif replied:
can also go into `tests/kernel/poll/`, `tests/integration/kernel` is not the right place, we need this per module/kernel feature and all in one place related to the primitive being changed.

TaiJuWu replied:
Done.

nashif reviewed (changes requested):
please put tests in the right place.

TaiJuWu commented:
Hi @peter-mitsis ,

Let me share the performace impact for this change.
This benchmark comes from `tests/benchmarks/app_kernel`

With this PR
|Method|       time |
|------------------------------------------------------------------------|----------|
| enqueue 1 byte msg in MSGQ                                                        |      7362|
| enqueue 4 bytes msg in MSGQ                                                      |      8514|
| enqueue 192 bytes msg in MSGQ                                                  |     80706|
| enqueue 1 byte msg in MSGQ to a waiting higher priority task    |     34498|
| enqueue 4 bytes in MSGQ to a waiting higher priority task          |     35650|
| enqueue 192 bytes in MSGQ to a waiting higher priority task      |    107841



Without this PR
|Method|       time |
|------------------------------------------------------------------------|----------|
| enqueue 1 byte msg in MSGQ                                       |      6530|
| enqueue 4 bytes msg in MSGQ                                      |      7682|
| enqueue 192 bytes msg in MSGQ                                    |     79874|
| enqueue 1 byte msg in MSGQ to a waiting higher priority task     |     34434|
| enqueue 4 bytes in MSGQ to a waiting higher priority task        |     35586|
| enqueue 192 bytes in MSGQ to a waiting higher priority task      |    107777|

The performance is not significantly affected when the message size exceeds 192B.
Do we still need to conditional `z_reschedule()` here since I also think it is tricky thing for condition `z_reschedule()`.
I also can image there will similiar issue for the old implementation (https://github.com/zephyrproject-rtos/zephyr/commit/80cd1f933d3a1779122a37063e4bd88ae343f093).

I can image below
1. Thread A is created with high priority
2. Thread B put message but there is no thread on waiting queue
3. Thread B continue to work.

andyross commented:
> I don't think that an unconditional z_reschedule() is the right approach; it seems too heavy handed for when there is an uncontested resource

reschedule() doesn't call down directly into swap, it has its own test of the run queue to see if something needs to happen.   It should be cheap and if it's missing a case we should fix that.

TaiJuWu commented:
Update: only move test into `test/kernel/msgq`

TaiJuWu committed:
test: k_msgq_put should trigger reschedule

In this test, low priority thread call `k_msgq_put` should
trigger reschedule.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

andyross reviewed (approved):
Refresh +1

mmahadevan108 commented:
@peter-mitsis, can you please re-review
